### ISSUE TITLE
Low precision inference?

### ISSUE THREAD

**Issue by akshatd** [pjreddie/darknet#81]
Hi!
Are there plans for making a low precision inference mode like many other neural network frameworks out there?
Would be really helpful for embedded applications where we have very limited memory!

**Comment by adroit91** [comment]
There is "binary" and "xnor" mentioned internally in the convolutional layer files. If you can make them to work, that is.

**Comment by akshatd** [comment]
I was thinking more on the level of the whole thing being reduced to 8 bit like in tensorflow, with weight conversion and 8-bit operators for each layer:  [https://www.tensorflow.org/performance/quantization](url).
I think this is a very natural path for this framework, as almost everyone is trying to do this to extract more performance out of their hardware.
If I had to do this myself, could anyone suggest how i could start?

**Comment by adroit91** [comment]
The correct positions where @pjreddie does binarization is probably where you could insert your quantization instead, and then proceed with GEMM, etc. You will have to write your own low precision version of the same functions, as the current code isn't templated. Recommended checking out libraries like tinyDNN to see if they already have some implementation for the same.

**Comment by sivagnanamn** [comment]
@adroit91 How to use binary and xnor options while training? I added xnor=1 and binary=1 in train cfg and started training, but Avg IOU and other parameters were always 0.

The model got saved after 100 iterations but there was no reduction in size of the model. It'll be helpful if you could share sample cfg

**Comment by TaihuLight** [comment]
In the Darknet Google Group(),the xnor_layer was removed at some point (https://github.com/pjreddie/darknet/commit/32d2c969973aa98635123743f321859192ff581d).The code at point (https://github.com/pjreddie/darknet/tree/8a6ba2fff3ee1c14bca0aa0e0a909aba7057cc94) can be complied and run successfully according to the users of the Group, but i can do it.

**Comment by zrobotparking** [comment]
@TaihuLight  thanks for the information.
I try to train `xyolo.test.cfg` with following code but can't finish training.
> ./darknet yolo train cfg/xyolo.test.cfg darknet.conv.weights 

The training start but the result turn into a huge number and fail.
Same situation on CPU and GPU training.
![image](https://user-images.githubusercontent.com/11483741/39958525-d7d1efea-5636-11e8-81fb-7a2b536ec359.png)


Buy the way, what is the purpose for the following code?
> make -f ai2.mk
Is it the implementation for binary conv. ?

**Comment by AlexeyAB** [comment]
@zrobotparking 
Try to train this XNOR-model: [tiny-yolo-obj_xnor.zip](https://github.com/pjreddie/darknet/files/1997701/tiny-yolo-obj_xnor.zip)
as common yolov2-tiny.cfg:
`./darknet detector train data/obj.data tiny-yolo-obj_xnor.cfg`
or
`./darknet detector train data/obj.data tiny-yolo-obj_xnor.cfg tiny-yolo-voc.conv.13`

So 1st and last conv-layers should be FP32, and others XNOR=1 (1-bit).

**Comment by zrobotparking** [comment]
@AlexeyAB Thanks a lot!!
It successfully start training.
I train the model under your darknet fork.

**Comment by AlexeyAB** [comment]
@zrobotparking After ~2000 iterations you will can successfully detect your objects using XNOR-model.

Also you should know, that optimized XNOR binary inference was removed from the Darknet. So although there is used 1-bit inference, it is executed using float-arithmetic, so detection will be slow. You should implement your own XNOR-backend for `conv, maxpool, route, reorg, upsample` layers. Or wait for somebody, who will make it :)

**Comment by zrobotparking** [comment]
@AlexeyAB  
I tried to `train tiny-yolo-obj_xnor.cfg` on VOC & COCO. But after 2000 iterations, the loss increase and become  nan. Same situation if I add `xnor=1` to convolution layer in yolov3-tiny.cfg .
If I use original tiny-yolo or tiny-yolov3 the training works perfect.

The cmd I used:
`
./darknet detector train data/voc.data tiny-yolo-obj_xnor.cfg
./darknet detector train data/voc.data yolov3-tiny-xnor.cfg

`

**Comment by AlexeyAB** [comment]
@zrobotparking 

Did you change `classes=20` and `filters=125` and `anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52` to train on Pascal VOC?
You should use such cfg-file for training on VOC - try to use it: [tiny-yolo-voc_xnor.zip](https://github.com/pjreddie/darknet/files/1998472/tiny-yolo-voc_xnor.zip)



Try to get pre-trained weights as here: https://github.com/AlexeyAB/darknet/blob/cda8171feb76bcb405350fd8341d42a0300e2f4b/build/darknet/x64/partial.cmd#L9
`./darknet partial cfg/yolov2-tiny-voc.cfg yolov2-tiny-voc.weights yolov2-tiny-voc.conv.13 13`

And train
`./darknet detector train data/obj.data tiny-yolo-voc_xnor.cfg tiny-yolo-voc.conv.13`

I didn't train it on Pascal VOC, but I trained it on my custom dataset, and it works with ~15% of mAP decreasing only.

If you will still get increasing of loss, try to use this GitHub repo: https://github.com/AlexeyAB/darknet

**Comment by sunshinezhihuo** [comment]
@AlexeyAB  Hello, 
① I just want to detect "person", that is, only one class.  So I use person_val.txt in VOC datasets. And I set classes =1 in cfg/voc.data and set person in data/voc.names. Then I set classes =1 and some filters = 18 in cfg/yolov3.cfg.  Do you think my method is correct?  About background, is there need to add to classes? That is , classes includes ["person","background"] and set classes = 2.  
② In yolov3.cfg, I set:
`batch =128  subdivisions =64  learning_rate = 0.0001.`
When I train the network, the loss decreases at the beginning. After some iterations, the loss becomes increased and then decreased . Should I change the learning_rate more smaller? 
Thank you for your reply.

**Comment by AlexeyAB** [comment]
@zrobotparking 
I fixed it in my repository: https://github.com/AlexeyAB/darknet

**Comment by Thilanka97** [comment]
Do you guys know of any Binarized yolov3 implementation ? where I can get pre-trained weights for binarized yolov3 ? or would you guide me to implement binarized yolov3 ?

**Comment by AlexeyAB** [comment]
@Thilanka97 There is implemented acceleration on XNOR: https://github.com/AlexeyAB/darknet
More info: https://github.com/AlexeyAB/darknet/issues/1472

**Comment by phonygene** [comment]
Guys ,  I add xnor = 1 under [convolutuinal] tag in .cfg as example.
But no matter which I used , xnor or binary , the . weights files size didn't get smaller .
And the testing time was even increase by four times.
Isn't it supposed to be faster than before ? 

I only changed convolutional layers before yolo layers like below :
```
[convolutional]
binary = 1
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky
```

**Comment by Thilanka97** [comment]
@AlexeyAB hey, if we change the cfg file to xnor and train, then it will train yolov2 using binary weights and will return binary weights right? Am I right here ? I just want to know if the yolov2 xnor code is available. I mean an inplementation of yolov2(tiny) using xnor-net or binary BNN? 

Thanks in advance!
